src/librbd/io/ImageRequestWQ.cc

```diff
@@ -576,25 +581,89 @@ void ImageRequestWQ<I>::set_require_lock(Direction direction, bool enabled) {
   }
 }
 
+template <typename I>
+void ImageRequestWQ<I>::set_qos_limit(uint64_t limit, const uint64_t flag) {
+  TokenBucketThrottle *throttle = nullptr;
+  for (auto pair : m_throttles) {
+    if (flag == pair.first) {
+      throttle = pair.second;
+      break;
+    }
+  }
+  assert(throttle != nullptr);
+  throttle->set_max(limit);
+  throttle->set_average(limit);
+  if (limit)
+    m_qos_enabled_flag |= flag;
+  else
+    m_qos_enabled_flag &= ~flag;
+}
+
 template <typename I>
 void ImageRequestWQ<I>::apply_qos_iops_limit(uint64_t limit) {
-  iops_throttle->set_max(limit);
-  iops_throttle->set_average(limit);
+  set_qos_limit(limit, RBD_QOS_IOPS_THROTTLE);
+}
+
+template <typename I>
+void ImageRequestWQ<I>::apply_qos_bps_limit(uint64_t limit) {
+  set_qos_limit(limit, RBD_QOS_BPS_THROTTLE);
+}
+
+template <typename I>
+void ImageRequestWQ<I>::apply_qos_read_iops_limit(uint64_t limit) {
+  set_qos_limit(limit, RBD_QOS_READ_IOPS_THROTTLE);
+}
+
+template <typename I>
+void ImageRequestWQ<I>::apply_qos_write_iops_limit(uint64_t limit) {
+  set_qos_limit(limit, RBD_QOS_WRITE_IOPS_THROTTLE);
 }
 
 template <typename I>
-void ImageRequestWQ<I>::handle_iops_throttle_ready(
-    int r, ImageDispatchSpec<I> *item) {
+void ImageRequestWQ<I>::apply_qos_read_bps_limit(uint64_t limit) {
+  set_qos_limit(limit, RBD_QOS_READ_BPS_THROTTLE);
+}
+
+template <typename I>
+void ImageRequestWQ<I>::apply_qos_write_bps_limit(uint64_t limit) {
+  set_qos_limit(limit, RBD_QOS_WRITE_BPS_THROTTLE);
+}
+
+template <typename I>
+void ImageRequestWQ<I>::handle_throttle_ready(int r, ImageDispatchSpec<I> *item, uint64_t flag) {
   CephContext *cct = m_image_ctx.cct;
   ldout(cct, 15) << "r=" << r << ", " << "req=" << item << dendl;
 
-  assert(m_io_blockers.load() > 0);
-  --m_io_blockers;
-  item->set_throttled();
+  assert(m_io_throttled.load() > 0);
+  --m_io_throttled;
+  item->set_throttled(flag);
   this->requeue(item);
   this->signal();
 }
 
+template <typename I>
+bool ImageRequestWQ<I>::needs_throttle(ImageDispatchSpec<I> *item) { 
+  uint64_t tokens = 0;
+  uint64_t flag = 0;
+  TokenBucketThrottle* throttle = nullptr;
+
+  for (auto t : m_throttles) {
+    flag = t.first;
+    if (!(m_qos_enabled_flag & flag))
+      continue;
+
+    throttle = t.second;
+    tokens = item->tokens_requested(flag);
+    if (!item->was_throttled(flag) &&
+	  throttle->get<ImageRequestWQ<I>, ImageDispatchSpec<I>,
+	      &ImageRequestWQ<I>::handle_throttle_ready>(
+		tokens, this, item, flag)) {
+      return true;
```

Comment:
Should the item by applied to all possible throttles in the first-pass so that it doesn't enter the back of the line when it's requeued? Perhaps the ```ImageDispatchSpec::was_throttled```/```ImageDispatchSpec::set_throttled``` logic should be reversed so that you default to all throttle bits set and you clear them if the throttle isn't required or when the throttle expires in ```handle_throttle_ready```

Reply:
Sorry I am afraid I did not catch all your point here :( .

But I did not set the all throttle bits for each request, instead, only check the throttles with m_qos_enabled_flag, because I consider this route is a hot route in IO path, I wish we can use as less time as possible here.

Could you help to explain more with your comment? Thanx a lot.

Reply:
You need to avoid a race where a throttle completes while you are still iterating through throttles so that's why I was suggesting that it might be either to invert the default initialization state of the ```was/set_throttled``` bitmask (i.e. start w/ the assumption that throttling was required and turn it off bit-by-bit if it wasn't).

Reply:
Hmm, but the needs_throttle() and handle_throttle_ready() can't be called at the same time. I mean, we call needs_throttle() in _void_dequeue() when the request is in first-pass or requeued, the request is still in queue. But the handle_throttle_ready() is called when the request is already dequeued and throttled. So I think there is no race between them.

I think this is not what you want to say, I still did not get your point, sorry for that. could you please give me an example for it. Thanx a lot. :)

Reply:
Yup -- you have two throttles active (rate and IOPS) and the request is throttled by both. The ```handle_throttle_ready``` will be called twice for the same request.

Reply:
Yes,  it would be called twice, but there is a flag to specify what we need to handle in handle_throttle_ready(), just something like that:
```
we have an image test with bps and iops throttle.


ImageRequestWQ:                                                                 BPS_Throttle                                    IOPS_Throttle
        |
        |---_void_dequeue()
                |
                |--needs_throttle()
                        |
                        |--throttle->get() <flag is for bps>         ---->   void TokenBucketThrottle::add_tokens()
                                                                                |
                                                                                |--b.ctx->complete(0);
        |
        |--handle_throttle_ready(flag) <flag is for bps>             <----
                |
                |---requeue(request)

        |
        |--_void_dequeue()
                |
                |--needs_throttle()
                        |
                        |--throttle->get() <flag is for iops>                                                   ---->   void TokenBucketThrottle::add_tokens()
                                                                                                                                |
                                                                                                                                |--b.ctx->complete(0);
        |
        |--handle_throttle_ready(flag) <flag is for iops>                                                       <----
                |
                |---requeue(request)

        |
        |--_void_dequeue()
                |
                |--needs_throttle() <no need to throttle now>
```

But, BTW, I found another problem here, if we get enough tokens, I forget to set_throttled(). so there is a fix as below:
```
diff --git a/src/librbd/io/ImageRequestWQ.cc b/src/librbd/io/ImageRequestWQ.cc
index dad7b2a..bbca232 100644
--- a/src/librbd/io/ImageRequestWQ.cc
+++ b/src/librbd/io/ImageRequestWQ.cc
@@ -654,11 +654,16 @@ bool ImageRequestWQ<I>::needs_throttle(ImageDispatchSpec<I> *item) {
 
     throttle = t.second;
     tokens = item->tokens_requested(flag);
-    if (!item->was_throttled(flag) &&
-         throttle->get<ImageRequestWQ<I>, ImageDispatchSpec<I>,
+
+    if (item->was_throttled(flag))
+      continue;
+
+    if (throttle->get<ImageRequestWQ<I>, ImageDispatchSpec<I>,
              &ImageRequestWQ<I>::handle_throttle_ready>(
                tokens, this, item, flag)) {
       return true;
+    } else {
+      item->set_throttled(flag);
     }
   }
   return false;
```

Reply:
.... but getting back to the original comment, I think you should apply all applicable throttles in the initial pass so that a request that needs to be throttled by more than one throttle doesn't need to wait at the back of the second (or third ...) throttle queue after it's requeued.